### PR TITLE
Added Mistral

### PR DESCRIPTION
**Pull request by kimbochen**
Added Mistral 7B that supports sliding window attention.

## Performance

```
python train.py configs/mistral-v0.1-proxy.json --bsz 2
After 32 Warmup: Mean TFLOP/s: 162.62 Mean MFU: 16.43%
``` 

## FLOPs per Token Estimation

Density = Sliding window size / Max sequence length.  
In `Mistral 7B v0.1`'s case, density = 4096 / 32768 = 0.125.

## Sliding Window Attention Mask Example

```
create_sliding_window_attention_mask(3, 10):
tensor([
  [ True, False, False, False, False, False, False, False, False, False],                                                                                   [ True,  True, False, False, False, False, False, False, False, False],
  [ True,  True,  True, False, False, False, False, False, False, False],
  [False,  True,  True,  True, False, False, False, False, False, False],
  [False, False,  True,  True,  True, False, False, False, False, False],
  [False, False, False,  True,  True,  True, False, False, False, False],
  [False, False, False, False,  True,  True,  True, False, False, False],
  [False, False, False, False, False,  True,  True,  True, False, False],
  [False, False, False, False, False, False,  True,  True,  True, False],
  [False, False, False, False, False, False, False,  True,  True,  True]
])
```

### CODE REVIEW

**Review by OrenLeung — COMMENTED** [review]
Thanks for the PR!

I was wondering if you can also do fp8. Should be fairly straightforward as te.TransformerLayer supports swa Out of the box.

And can you double check using the flex attention get sparsity API that it is 12.5%?

Furthermore, Can you also implement SWA using flex attention 

If nvidia
   Use flex attention
Eif amd
   Use sdpa + mask

**Comment by kimbochen** [comment]
## Flex attention selector (if Nvidia use flex attention, sdpa otherwise)

`enable_gqa=True` not used due to my PyTorch version not new enough.

Performance:
`python train.py configs/mistral-v0.1-proxy.json --bsz 2`
- SDPA: Mean TFLOP/s: 336.45 Mean MFU: 34.00%
- Flex Attention: Mean TFLOP/s: 574.68 Mean MFU: 58.08%

## Native TE sliding window

`python train.py configs/mistral-v0.1-proxy.json --bsz 2 --use-fp8`
Performance:
- Custom SWA mask: Mean TFLOP/s: 1001.77 Mean MFU: 50.62%
- Native SWA mask: Mean TFLOP/s: 1119.14 Mean MFU: 56.55%

**Review by OrenLeung — APPROVED** [review]
LGTM, nicely done! 🥳 

can you resolve the merge conflict then we can merge?

**Comment by OrenLeung** [comment]
also, can you post the results of the fully compiled model swa vs full compiled model sdpa + mask?

**Comment by kimbochen** [comment]
> also, can you post the results of the fully compiled model swa vs full compiled model sdpa + mask?

`python train.py configs/mistral-v0.1-proxy.json --bsz 2 --pt-compile`

- Flex Attention: Mean TFLOP/s: 714.93 Mean MFU: 72.25%
- SDPA: Mean TFLOP/s: 383.25 Mean MFU: 38.73%
